### ISSUE TITLE
Sort the values of an nD array

### ISSUE THREAD

**Issue by slevang** [pydata/xarray#8129]
### Is your feature request related to a problem?

As far as I know, there is no straightforward API in xarray to do what `np.sort` or `pandas.sort_values` does. We have `DataArray.sortby("x")`, which will sort the array according to the coordinate itself. But if instead you want to sort the values of the array to be monotonic, you're on your own. There are probably a lot of ways we _could_ do this, but I ended up with the couple line solution below after a little trial and error.

### Describe the solution you'd like

Would there be interest in implementing a `Dataset`/`DataArray.sort_values(dim="x")` method?

_Note: this 1D example is not really relevant, see the 2D version and more obvious implementation in comments below for what I really want._

```python
def sort_values(self, dim: str):
    sort_idx = self.argsort(axis=self.get_axis_num(dim)).drop_vars(dim)
    return self.isel({dim: sort_idx}).drop_vars(dim).assign_coords({dim: self[dim]})
```

The goal is to handle arrays that we want to monotize like so:

```python
da = xr.DataArray([1, 3, 2, 4], coords={"x": [1, 2, 3, 4]})
da.sort_values("x")

<xarray.DataArray (x: 4)>
array([1, 2, 3, 4])
Coordinates:
  * x        (x) int64 1 2 3 4
```
![image](https://github.com/pydata/xarray/assets/39069044/216764fa-1f03-4db4-acfd-86af4abfc0ea)

In addition to `sortby` which can deal with an array that is just unordered according to the coordinate:

```python
da = xr.DataArray([1, 3, 2, 4], coords={"x": [1, 3, 2, 4]})
da.sortby("x")

<xarray.DataArray (x: 4)>
array([1, 2, 3, 4])
Coordinates:
  * x        (x) int64 1 2 3 4
```

### Describe alternatives you've considered

I don't know if `argsort` is dask-enabled (the docs just point to the `numpy` function). Is there a more intelligent way to implement this with `apply_ufunc` and something else? I assume chunking in the sort dimension would be problematic.

### Additional context

Some past related threads on this topic:
https://github.com/pydata/xarray/issues/3957
https://stackoverflow.com/questions/64518239/sorting-dataset-along-axis-with-dask

**Comment by benbovy** [comment]
According the [the docs](https://docs.xarray.dev/en/latest/generated/xarray.DataArray.sortby.html), you could already sort by array values by passing a DataArray:

```python
da = xr.DataArray([1, 3, 2, 4], coords={"x": [1, 2, 3, 4]})

da.sortby(da)
# <xarray.DataArray (x: 4)>
# array([1, 2, 3, 4])
# Coordinates:
#   * x        (x) int64 1 3 2 4
```

Is this what you want to achieve? Or do you rather want to sort *only* the array values without reordering the coordinates (i.e., `da.sortby(da).assign_coords(x=da.x)`)? Or maybe you want to support sorting a nD array by values along a given dimension?

**Comment by slevang** [comment]
Yes `nD` arrays is actually the use case I have in mind here, I just did a 1D example for simplicity but your suggestion is much better there. Is there a way to achieve that with the current `sortby`? I think not but would be great if there was. I guess the key feature of `pd.sort_values` is it has an `axis` arg, which I would like to generalize to `nD` xarray objects.

**Comment by benbovy** [comment]
> Is there a way to achieve that with the current sortby?

I don't think either.

One problem with sorting a nD DataArray by value along dimension `dim` is how to treat its coordinates? Re-ordering coordinate labels would be ambiguous when the other dimensions have a size > 1. Not re-ordering the labels at all would make them out-of-sync with the data. Should we drop the coordinates in this case? Or add a new method to `xarray.Variable` instead?

> I guess the key feature of pd.sort_values is it has an axis arg, which I would like to generalize to nD xarray objects.

I'm not sure to fully understand this. Considering `da.sortby(coord_name_or_dataarray)`:

- the values/labels of `coord_name_or_dataarray` corresponds to the `by` argument of pd.sort_values
- `coord_name_or_dataarray` must either be or refer to a 1-dimensional DataArray, and its dimension corresponds to the `axis` argument of pd.sort_values

**Comment by benbovy** [comment]
A 2-d example of array sort by value using `sortby`:

```python
da2 = xr.DataArray(
    [[4, 3, 2, 1], [2, 3, 1, 4]],
    coords={"x": [1, 2, 3, 4]},
    dims=("y", "x")),
)
# <xarray.DataArray (y: 2, x: 4)>
# array([[4, 3, 2, 1],
#        [2, 3, 1, 4]])
# Coordinates:
#   * x        (x) int64 1 2 3 4
# Dimensions without coordinates: y

da2.sortby(da2.isel(y=1))
# <xarray.DataArray (y: 2, x: 4)>
# array([[2, 4, 3, 1],
#        [1, 2, 3, 4]])
# Coordinates:
#   * x        (x) int64 3 1 2 4
# Dimensions without coordinates: y
```

**Comment by slevang** [comment]
It appears there is [no chunk-aware version](https://github.com/dask/dask/issues/4368) of this implemented in `dask` so using `np.sort` is probably the best we can do.

**Comment by benbovy** [comment]
I understand better, thanks. Yes this is something different and actually simpler than a generalization of pd.sort_values.

I think it would better fit as part of the `xarray.Variable` API (or the `NamedArray` API #8075) since coordinate labels are irrelevant here.

**Comment by slevang** [comment]
Yep makes sense. If I were to get around to opening a PR for this, what do you think the appropriate naming would be? 

1. `sort_values` is descriptive, but you're right that this doesn't quite duplicate the functionality of `pd.sort_values`, which can also bring the labels along with the values
2. `sort` would match the lower-level call since this is really `np.sort`, but I think having `DataArray.sort` and `DataArray.sortby` could be pretty confusing
3. we could add a `dim` argument to `sortby` that would handle this task, but that could also be confusing in pushing too much different functionality into one method

**Comment by benbovy** [comment]
I like `xr.Variable.sort`, which is consistent with `np.sort`. I wouldn't mind having both `xr.Variable.sort` and `xr.DataArray.sortby`. Maybe it would be better to wait for #8075 before adding any API to Variable?

I wouldn't add `DataArray.sort` or `Dataset.sort` for the reasons that I mentioned in my previous comment. To have the sorted array wrapped in a DataArray, I think I'd prefer an API like:

```python
result = xr.DataArray(da.variable.sort("x"))

# or if one wants the coordinates too

result = xr.DataArray(da.variable.sort("x"), coords=da.coords)
```

Admittedly It is a bit more verbose, but it is more explicit regarding the fact that `sort` has nothing to do with the coordinate labels.

**Comment by slevang** [comment]
That's a reasonable approach, although I think the `Variable` API is much less well known to the average user. This would also get quite verbose for `Dataset`:
```python
result = xr.Dataset({v: xr.DataArray(ds[v].variable.sort("x"), coords=ds[v].coords) for v in ds.data_vars})
```

I won't get to this anytime soon, so I'm happy to wait for #8075 and leave this open for any other input on the design. In the meantime, `apply_ufunc(np.sort, ...)` is the easiest way to handle this for anyone else with the same question.

**Comment by benbovy** [comment]
> although I think the Variable API is much less well known to the average user.

Fair enough, although the goal of #8075 is to bring it more in the spotlight (as a `NamedArray`).

Note that you can use `Dataset.map`:

```python
result = ds.map(lambda da: da.variable.sort("x"))
```
